ENT-97: Email learners on enrollment

haikuginger opened the pull request:
**Description:**

This pull request adds a feature that enables platform owners to email new learners upon enrollment in a course, using either a built-in template with approved text, or by setting up a custom, site-specific, set of templates that can be used to render messages.

**JIRA:** [ENT-97](https://openedx.atlassian.net/browse/ENT-97)

**Dependencies:** Depends on an incoming PR from @e-kolpakov; the ticket acceptance criteria specify that learners can be notified when being enrolled in a program. However, program enrollment does not exist as of yet.

**Merge deadline:** 30 January 2017

**Sandbox Details:** TBD (being provisioned)

**Testing instructions:**

1. In the Manage Learners page in a devstack (where emails are logged to the console), add a learner to a course using an arbitrary email address.
    - Observe in the console that an email has been "sent" using both the standard plaintext template and the standard HTML template - email clients will choose the appropriate one to render.
2. Set up an `EnrollmentNotificationEmailTemplate` object in the Enterprise admin; be sure to fill in templates for both the plaintext and HTML templates, as well as for the subject line.
3. Again, add a learner to a course using an arbitrary email address.
    - Observe in the console that the email was "sent", and that your custom templates were used to render the message, rather than the standard templates.

**Reviewers:**
- [x] @bradenmacdonald 
- [ ] edX TBD

**Merge checklist:**
- [x] All reviewers approved
- [x] CI build is green
- [x] Version bumped
- [x] Changelog record added
- [x] Commits are (reasonably) squashed
- [x] Translations are updated
- [x] PR author is listed in AUTHORS

**Post merge:**
- [ ] Create a tag
- [ ] Check new version is pushed to PyPi after tag-triggered build is 
      finished.
- [ ] Delete working branch (if not needed anymore)

**Author concerns:**

1. More tests are still incoming; this is not entirely complete yet.
2. A flag needs to be added to the Enterprise admin to make notification optional
3. Because program enrollment doesn't exist yet, notification on program enrollment isn't implemented yet.

edx-webhook commented:
Thanks for the pull request, @haikuginger! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-enterprise&number=27"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

haikuginger commented:
Rebased from caeb0ee.

bradenmacdonald commented on enterprise/models.py:
```diff
@@ -541,3 +543,71 @@ def __repr__(self):
+
+    SUBJECT_HELP_TEXT = _(
+        'Fill in a Python format string, using the same variables as are available for '
+        'the HTML and plaintext body template.'
```
Can you provide an example?

bradenmacdonald commented on enterprise/templates/enterprise/emails/user_notification.html:
```diff
@@ -0,0 +1,10 @@
+</p>
+<p>
+The {{course_name}} team
+</p>
```
An HTML email template should have a full `<html>` document - see http://localhost:8000/admin/bulk_email/courseemailtemplate/ on your devstack for an example. It would even be good to use that as a template here, since it's more detailed and pretty. (Arguably too detailed).

Also, admin users should have some way of seeing a default template - I think that's done somehow for http://localhost:8000/admin/bulk_email/courseemailtemplate/ but for this new template, I have no examples to work from in the django admin.

haikuginger replied:
Re: the content: that's prescribed by the ticket, so I'm not sure if we want a different template, especially as we're using a different set of data here.

But I agree that there should be a preview view.

bradenmacdonald replied:
At the very least, the template should have `<html>`, `<body>` etc. Check out https://github.com/leemunroe/responsive-html-email-template for an example that's very nice and also where the code is easy to read, unlike the edx example.

bradenmacdonald commented on enterprise/templates/enterprise/emails/user_notification.txt:
```diff
@@ -0,0 +1,9 @@
+{% if user_name %}Dear {{ user_name }},{% else %}Hi!{% endif %}
+
+You have been enrolled in {{ course_name }}, a {% if program_type %}{{ program_type }} {% endif %}{{ course_type }} offered by {{ organization_name }}. This {{ course_type }} begins {{ start_date | date:"F j, Y" }}. For more information, see the following link:
```
Can you mark this up for i18n?

Also, this way of combining integrating `course_type` and other vars into a sentence does not work for translation, even if the word `"course"` is translated, because it may need to be inflected differently depending on the context of the translated sentence. So you'll need to refactor this a bit.

```
{%if is_a_program%}
{% blocktrans %}
You've been enrolled in the {{name}} {{program_type}} program, ...
{% blocktrans %}
{%else%}
...
{%endif%}
```

(`program_type` can be inlined since it's a trademarked name and would not be translated nor inflected.)

haikuginger replied:
Makes sense.

bradenmacdonald commented on enterprise/utils.py:
```diff
@@ -166,3 +171,97 @@ def get_catalog_admin_url_template():
+    """
+    if hasattr(user, 'first_name') and hasattr(user, 'username'):
+        # PendingEnterpriseCustomerUsers don't have usernames or real names. We should
+        # template slightly differently to make sure weird stuff doesn't happen.
```
Needs to be mentioned in the admin help text or explained clearly in a comment in the example template. (Without referencing any python class names :-p )

haikuginger replied:
Agreed.

bradenmacdonald commented on enterprise/utils.py:
```diff
@@ -166,3 +171,97 @@ def get_catalog_admin_url_template():
+        'course_url': course_url
+    }
+    try:
+        current_site = get_current_site(request)
```
What happens if an edx staff person is logged in to the django admin on edx.org, but enrolls a learner in a WL site course? Wouldn't this send the edx email template, when we actually want the template of that particular WL site?

haikuginger replied:
Right now, it'd send from edx.org - good catch. Possibly grab the correct site off the EnterpriseCustomer?

bradenmacdonald replied:
Yep, grab the site off the EnterpriseCustomer - that sounds simple and would provide the correct result.

bradenmacdonald commented on enterprise/utils.py:
```diff
@@ -166,3 +171,97 @@ def get_catalog_admin_url_template():
+        if site_template.subject_line:
+            subject_line = site_template.subject_line.format(**msg_context)
+        else:
+            subject_line = 'You\'ve been enrolled in {}!'.format(course)
```
Optional: Can you just make `'You\'ve been enrolled in {course_name}!' the `default` on the model, so you don't have to have so many if/else conditions here?

haikuginger replied:
We won't always have the model available, but I'll take a look at what we can do.

bradenmacdonald commented on enterprise/utils.py:
```diff
@@ -166,3 +171,97 @@ def get_catalog_admin_url_template():
+            html_msg = render_to_string(
+                'enterprise/emails/user_notification.html',
+                msg_context
+            )
```
Can you refactor this so that the

```python
+            plain_msg = render_to_string(
 +                'enterprise/emails/user_notification.txt',
 +                msg_context
 +            )
 +            html_msg = render_to_string(
 +                'enterprise/emails/user_notification.html',
 +                msg_context
 +            )
```
code doesn't appear twice?

haikuginger replied:
Yes.

haikuginger commented on enterprise/admin/views.py:
```diff
@@ -273,6 +292,17 @@ def _enroll_users(cls, enterprise_customer, emails, course_id, mode, request):
+                        pending_user,
+                        course_details['name'],
+                        course_url,
+                        _('course'),
```
Makes sense.

mattdrayer replied:
Along the lines of readability, this private/internal function is orchestrating a fair amount of work and it might be better to break it up into additional private/internal functions.  I understand that it's possible to go too far with that approach, but it would help to separate things a bit more, imho.

haikuginger commented on enterprise/utils.py:
```diff
@@ -166,3 +171,97 @@ def get_catalog_admin_url_template():
+            to get details for the email
+        course: The name of the course or program that the user's been enrolled in
+        course_url: The URL to the LMS details page for the course or program.
+        course_type: The classification of the learning unit ("course" or "program")
```
I'll think on it. I agree that the current naming is confusing.

haikuginger commented on enterprise/admin/views.py:
```diff
@@ -328,7 +358,15 @@ def post(self, request, customer_uuid):
                 course_mode = manage_learners_form.cleaned_data[ManageLearnersForm.Fields.COURSE_MODE]
                 course_id = course_details['course_id']
-                self._enroll_users(enterprise_customer, linked_learners, course_id, course_mode, request)
+                notify = manage_learners_form.cleaned_data.get(ManageLearnersForm.Fields.NOTIFY) == 'email'
```
I'll check into it.

mattdrayer replied:
Ah, I made a similar comment above, suggesting something like `settings.DEFAULT_NOTIFICATION_MECHANISM`

haikuginger commented on enterprise/admin/__init__.py:
```diff
@@ -18,8 +18,9 @@
-    EnterpriseCustomerIdentityProvider, HistoricalUserDataSharingConsentAudit, PendingEnterpriseCustomerUser,
+    EnrollmentNotificationEmailTemplate, EnterpriseCustomer, EnterpriseCustomerUser,
+    EnterpriseCustomerBrandingConfiguration, EnterpriseCustomerIdentityProvider,
+    HistoricalUserDataSharingConsentAudit, PendingEnterpriseCustomerUser,
```
I absolutely agree, but it'd be impossible to make the `isort` linter accept that without making that change on all existing code - which might cause some issues to PRs in flight. I'd love to handle that as a separate PR though.

mattdrayer replied:
Please feel free to log ENT project tickets for any tech debt you identify -- these can be used as filler-jobs for when there are small gaps at the ends of sprints.

mattdrayer commented on enterprise/admin/forms.py:
```diff
@@ -46,6 +46,15 @@ class ManageLearnersForm(forms.Form):
+    notify_on_enrollment = forms.ChoiceField(
+        label=_("Notify learners of enrollment"),
+        choices=[
+            ("email", _("Email")),
```
Since it appears that we are describing the notification mechanism with these choices, we might instead say "by_email", instead of just "email".

mattdrayer commented on enterprise/admin/forms.py:
```diff
@@ -103,6 +113,12 @@ def clean_course(self):
+        """
+        Clean the notify_on_enrollment field.
+        """
+        return self.cleaned_data[self.Fields.NOTIFY] or 'email'
```
should `'email'` here be instead something like `settings.DEFAULT_NOTIFICATION_MECHANISM`?

mattdrayer commented on enterprise/admin/views.py:
```diff
@@ -218,7 +222,7 @@ def _handle_bulk_upload(cls, enterprise_customer, manage_learners_form, request)
 
     @classmethod
-    def _enroll_users(cls, enterprise_customer, emails, course_id, mode, request):
+    def _enroll_users(cls, enterprise_customer, emails, course_id, mode, request, notify=True):
```
Please add `notify` to the arguments listed in the docstring

mattdrayer commented on enterprise/admin/views.py:
```diff
@@ -234,23 +238,38 @@ def _enroll_users(cls, enterprise_customer, emails, course_id, mode, request):
-                    dict(user=username, message=error_message),
+                    dict(user=user.username, message=error_message),
                 )
             else:
```
For my own knowledge, why use `else` here versus using a `continue` in the preceding `except` block?

haikuginger replied:
To my knowledge, they'd work the same at the moment, but a `continue` might cause issues later on if we want to add something that happens in the loop, even after an exception, later on. I chose not to modify the existing code.

mattdrayer commented on enterprise/models.py:
```diff
@@ -541,3 +543,71 @@ def __repr__(self):
+
+    BODY_HELP_TEXT = _(
+        'Fill in a standard Django template; available variables include `user_name`, '
+        '`course_name`, `program_type`, `course_type`, `organization_name`, `start_date`, '
```
`program_type`, for that matter

mattdrayer commented on enterprise/models.py:
```diff
@@ -541,3 +543,71 @@ def __repr__(self):
+        """
+        return self.render_template(self.plaintext_template, kwargs)
+
+    def render_templates(self, kwargs):
```
If the goal of this function is to render ALL templates, then let's be explicit about that -> `render_all_templates()` -- otherwise I'd expect to be able to send in some sort of switch to control which template(s) are rendered.

haikuginger commented:
I've just pushed a big commit with a refactor of the code addressing the majority of the comments: Still remaining: set up a "preview" for the email template object, which I'll be doing either tonight or tomorrow.

bradenmacdonald commented on enterprise/admin/views.py:
```diff
@@ -229,28 +232,47 @@ def _enroll_users(cls, enterprise_customer, emails, course_id, mode, request):
         enrollment_client = EnrollmentApiClient()
+        course_details = CourseApiClient().get_course_details(course_id)
+        reverse_method = get_reverse_by_site(request, enterprise_customer.site)
+        course_url = reverse_method('about_course', args=(course_id,))
```
This `get_reverse_by_site` API seems needlessly complicated to me. Can't you just create a `reverse_site_url(request, site, *args, **kwargs)` method?

```python
course_url = reverse_site_url(request, site=enterprise_customer.site, view_name='about_course', args=(course_id, ))
```

You can also implement it simply as `return 'https://' + site.domain + reverse(*args, **kwargs)`.

Actually, on second thought, this assumes that the enterprise app and the LMS are deployed into the same django project. Does the Course catalog API not provide the URL of the course?

haikuginger replied:
> This get_reverse_by_site API seems needlessly complicated to me. Can't you just create a reverse_site_url(request, site, *args, **kwargs) method?

Certainly! I wanted to produce a method that had the same signature as the stock `reverse()` function, but I'll change it out.

haikuginger replied:
In re: your code suggestion. I'd rather not do it that way; it assumes that the site is accessible over https. My code here makes an assumption too; that the other site can be accessed using the same scheme as the admin site, but I think that's just enough safer that I'd rather keep it, even though it's more complicated.

> Actually, on second thought, this assumes that the enterprise app and the LMS are deployed into the same django project. Does the Course catalog API not provide the URL of the course?

As far as I can tell from available information, it does not. I've done extensive testing, and have been unable to locate anything that suggests it does.

bradenmacdonald replied:
> In re: your code suggestion. I'd rather not do it that way; it assumes that the site is accessible over https.

Yeah, but any production site should always be using HTTPS (and e.g. all OpenCraft and edX sandboxes do, and certainly anyone selling "enterprise" offerings would be using HTTPS). The only time HTTPs is not used AFAIK would be on devstacks, and then no emails are getting sent anyways - just logs to the console, so I assume it's not a big deal if the URLs shown in the console logged emails use the wrong protocol. But if that offends your sensibilities, it's fine to leave as is :)

bradenmacdonald replied:
@haikuginger I was talking to @mattdrayer just now and he said that the course discovery API should return a URL for each course, but only if the course is on the drupal marketing site. (So it would not currently work on devstack or on white label sites).

bradenmacdonald commented on enterprise/admin/views.py:
```diff
@@ -229,28 +232,47 @@ def _enroll_users(cls, enterprise_customer, emails, course_id, mode, request):
+        reverse_method = get_reverse_by_site(request, enterprise_customer.site)
+        course_url = reverse_method('about_course', args=(course_id,))
+        email_conn = mail.get_connection()
+        email_conn.open()
```
You are not closing the connection - better to use it as a context manager so you don't have to explicitly close it:

```python
with mail.get_connection() as connection:
```

Also, can we avoid opening the mail connection if we are not going to be notifying the users?

haikuginger replied:
Good catch! From what I can tell, though, those two items are mutually exclusive by default, so I'm putting together a context manager that conditionally opens and closes a connection.

bradenmacdonald commented on enterprise/admin/views.py:
```diff
@@ -229,28 +232,47 @@ def _enroll_users(cls, enterprise_customer, emails, course_id, mode, request):
+                            'start': datetime.datetime.strptime(course_details['start'], '%Y-%m-%dT%H:%M:%SZ'),
+                        },
+                        enterprise_customer=enterprise_customer,
+                        email_connection=email_conn,
```
This way of structuring the signature of this method is much nicer - thanks :)

bradenmacdonald commented on enterprise/admin/views.py:
```diff
@@ -229,28 +232,47 @@ def _enroll_users(cls, enterprise_customer, emails, course_id, mode, request):
+                            'name': course_details['name'],
+                            'url': course_url,
+                            'type': 'course',
+                            'start': datetime.datetime.strptime(course_details['start'], '%Y-%m-%dT%H:%M:%SZ'),
```
Is there some reason why so many edX API clients do not parse dates into python `datetime` objects? I don't like having date parsing logic scattered around all over the place.

mattdrayer replied:
I'm not sure why this is, but we could at least move the date-parsing logic into a utility method to contain the sprawl.

haikuginger replied:
The EdX API client is pretty sparse; it doesn't actually know much about the API structure. And, since JSON doesn't actually have a standard datetime format or way to specify that a field is, in fact, a datetime, there's not a reliable way to automatically parse a datetime from JSON.

bradenmacdonald commented on enterprise/admin/views.py:
```diff
@@ -328,7 +362,16 @@ def post(self, request, customer_uuid):
                 course_id = course_details['course_id']
-                self._enroll_users(enterprise_customer, linked_learners, course_id, course_mode, request)
+                notification_type = manage_learners_form.cleaned_data.get(ManageLearnersForm.Fields.NOTIFY)
+                notify = notification_type == ManageLearnersForm.NotificationTypes.BY_EMAIL
```
Are we expecting other types of notification in the future? Because in the form, you've defined the notification options as an enumerated choice set even though it's essentially a boolean choice, and in this part of the code you are explicitly treating it as a boolean.

mattdrayer replied:
I'd prefer to keep things open-ended and stick with the enum pattern -- I'm sure at some point someone will want Instagram support...

haikuginger replied:
I agree with Matt on this one.

bradenmacdonald replied:
Fine by me! Can we treat it as an enum everywhere though? Or maybe even a set :-p (e.g. why should "email" and "SMS", or whatever other options you envision be mutually exclusive). This change is not super important though imho.

bradenmacdonald commented on enterprise/templates/enterprise/emails/user_notification.html:
```diff
@@ -0,0 +1,15 @@
+<body>
+<p>{% if user_name %}{% blocktrans %}Dear {{ user_name }},{% endblocktrans %}{% else %}{% blocktrans %}Hi!{% endblocktrans %}{% endif %}</p>
+<p>{% if enrolled_in.type == "program" %}
+{% blocktrans with program_url=enrolled_in.url program_name=enrolled_in.name program_branding=enrolled_in.branding|pad_right start_date=enrolled_in.start|date:"F j, Y" %}You have been enrolled in <a href="{{ program_url }}">{{ program_name }}</a>, a {{ program_branding }}program offered by {{ organization_name }}. This program begins {{ start_date }}. For more information, see <a href="{{ program ...
```
For the HTML template, you don't need `pad_right` - just put a space in `{{ program_branding }}program`. Browsers automatically combine adjacent whitespace, so the end result is the same. I'd probably do the same for the plaintext too - it's simpler and worst case scenario is that some program with no branding name shows up as "a  program" and mode people won't even notice the extra space. Many modern mail clients render plain text emails using an HTML interface anyways, which means that adjacent spaces get combined there too.

bradenmacdonald replied:
For dates, it's better not to use any specific format like `F j, Y` since that is specific to English speakers. Instead, use  "DATE_FORMAT", "DATETIME_FORMAT", "SHORT_DATE_FORMAT" or "SHORT_DATETIME_FORMAT", which will pick an appropriate date format based on the active locale. ([ref](https://docs.djangoproject.com/en/1.10/ref/templates/builtins/#date))

e.g. `{{ value|date:"SHORT_DATE_FORMAT" }}`

haikuginger replied:
Ah, good to know! I actually didn't know about the automatic whitespace combination. I'll change to use the built-in date format string.

haikuginger commented:
@bradenmacdonald, I've just added a preview option for HTML email templates.

bradenmacdonald commented on enterprise/utils.py:
```diff
@@ -166,3 +173,156 @@ def get_catalog_admin_url_template():
+    The method is designed to fail in a "smart" way; if we can't render a
+    database-backed subject line template, then we'll fall back to a template
+    saved in the Django settings; if we can't render _that_ one, then we'll
+    fall through to a friendly string written into the code.
```
Can you mention why this would fall through?

haikuginger replied:
Will add - for future readers of this PR, one case might be if a site admin entered a subject line template with template tags other than `{course_name}`; that'd result in a `KeyError`.

bradenmacdonald commented:
Thanks @haikuginger  - looking good. Are you going to have a sandbox or should I plan to test locally?

mattdrayer commented on enterprise/admin/views.py:
```diff
@@ -12,23 +13,71 @@
+        template = get_object_or_404(EnrollmentNotificationEmailTemplate, pk=template_id)
+        if view_type not in self.view_type_contexts:
+            return HttpResponse(status=404)
+        logger.warning(template)
```
Just checking -- is this `logger` needed?

haikuginger replied:
Nope!

mattdrayer commented on enterprise/utils.py:
```diff
@@ -166,3 +173,156 @@ def get_catalog_admin_url_template():
+        return stock_subject_template.format(course_name=course_name)
+
+
+def send_notification_message(user, enrolled_in, enterprise_customer, email_connection=None):
```
If this method is solely-intended to handle email communications we should probably be more specific with the name -- eg `send_email_notification_message`

haikuginger replied:
Works for me!

haikuginger commented:
@bradenmacdonald, ready for another pass. I'm spawning a sandbox for testing, etc., but if you want to review in the meantime, the local setup should be extremely simple.

bradenmacdonald commented on enterprise/utils.py:
```diff
@@ -49,6 +49,39 @@ def __init__(self, *args, **kwargs):
+        """
+        Close the email connection if it was opened.
+        """
+        if self.conn and self.open_conn:
```
This is a nit, but it's better just to check `if self.conn` here - regardless of whether a connection was supposed to be open or not, if the connection was opened, it should be closed.

haikuginger replied:
Yeah, sounds good.

bradenmacdonald commented:
@haikuginger Thanks! Looking good to me. I appreciate the ability to preview templates from the admin - very handy.

:+1:

- [x] I tested this: tested on my devstack, saw emails coming through to the console.
- [x] I read through the code
- [x] I checked for accessibility issues
- [x] Includes documentation: somewhat - inline in the admin, yes.

haikuginger commented:
Rebased from 6666ae2.

@bradenmacdonald, I've added notification handling for Programs, now that @e-kolpakov's work is merged, and after doing some research, I've also moved some data gathering over to the course discovery API.

haikuginger commented:
@mattdrayer, I've squashed from b33b501; this is ready for merge whenever you declare it so.

mattdrayer commented:
👍 🚀

bradenmacdonald commented:
@haikuginger, FYI when verifying this on a sandbox without the catalog service, I got this stacktrace - might be worth providing better error handling if we can add that to a future PR?

```
Traceback (most recent call last):
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/handlers/base.py", line 132, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 145, in inner
    return func(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/newrelic-2.74.0.54/newrelic/hooks/framework_django.py", line 503, in wrapper
    return wrapped(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/utils/decorators.py", line 110, in _wrapped_view
    response = view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/decorators/cache.py", line 57, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/contrib/admin/sites.py", line 233, in inner
    return view(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/generic/base.py", line 71, in view
    return self.dispatch(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/newrelic-2.74.0.54/newrelic/hooks/framework_django.py", line 885, in wrapper
    return wrapped(*args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/views/generic/base.py", line 89, in dispatch
    return handler(request, *args, **kwargs)
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/enterprise/admin/views.py", line 471, in post
    notify=notify,
  File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/enterprise/admin/views.py", line 289, in _enroll_users
    course_url = course_details.get('marketing_url')
AttributeError: 'list' object has no attribute 'get'
```

mattdrayer commented:
@bradenmacdonald @haikuginger can we file a bug on this so the issue does not get lost in the shuffle?

haikuginger commented:
@mattdrayer, creating a story now.
